### DISCUSSION TITLE
Error in oidc_util_create_symmetric_key

### DISCUSSION THREAD

**Discussion by wetzel-desy** [OpenIDC/mod_auth_openidc#529] [Q&A] [answered]
Hi, 

I am running mod_auth_openidc and today encountered the following error during authentication:

src/util.c(2169): [client XXX] oidc_util_create_symmetric_key: key_len=30 free(): invalid pointer

Before I did a system update yesterday, the module worked as intended, has this happened to anyone else?
If you nee more info, please let me know.

Cheers and thanks,
Tim

**Comment by zandbelt** [comment]
which version are you on? also, please provide full server debug logs

**Reply by wetzel-desy** [reply]
Thanks for the quick reply.
I am running the latest version of mod_auth_oidc from the master branch with Apache/2.4.41 on Ubuntu 20.04.1 LTS.
I should add that I am using `AuthType oauth20`, which I know is deprecated but it has worked well for me. If that is a problem and mod_oauth2 would be a better fit, please let me know. Anyhow, I thought I should reported the issue I encountered.

Thanks.

Server logs (redacted bearer token and IP):
```
[Thu Jan 28 12:56:43.401855 2021] [authz_core:debug] [pid 1415] mod_authz_core.c(817): [client X.X.X.X:29452] AH01626: authorization result of <RequireAny>: denied (no authenticated user yet)
[Thu Jan 28 12:56:43.401886 2021] [auth_openidc:debug] [pid 1415] src/mod_auth_openidc.c(3838): [client X.X.X.X:29452] oidc_check_user_id: incoming request: "/serve?(null)", ap_is_initial_req(r)=1
[Thu Jan 28 12:56:43.401911 2021] [auth_openidc:debug] [pid 1415] src/util.c(1268): [client X.X.X.X:29452] oidc_util_request_matches_url: comparing "/serve"=="/oidc-redirect"
[Thu Jan 28 12:56:43.401920 2021] [auth_openidc:debug] [pid 1415] src/oauth.c(201): [client X.X.X.X:29452] oidc_oauth_get_bearer_token: accept_token_in=0
[Thu Jan 28 12:56:43.401930 2021] [auth_openidc:debug] [pid 1415] src/util.c(2403): [client X.X.X.X:29452] oidc_util_hdr_in_get: Authorization=Bearer [REDACTED]
[Thu Jan 28 12:56:43.401938 2021] [auth_openidc:debug] [pid 1415] src/oauth.c(215): [client X.X.X.X:29452] oidc_oauth_get_bearer_token: authorization header found
[Thu Jan 28 12:56:43.401951 2021] [auth_openidc:debug] [pid 1415] src/oauth.c(297): [client X.X.X.X:29452] oidc_oauth_get_bearer_token: bearer token: [REDACTED]
[Thu Jan 28 12:56:43.401960 2021] [auth_openidc:debug] [pid 1415] src/oauth.c(579): [client X.X.X.X:29452] oidc_oauth_validate_jwt_access_token: enter: JWT access_token header={"typ":"at+jwt","alg":"RS256"}
[Thu Jan 28 12:56:43.401968 2021] [auth_openidc:debug] [pid 1415] src/util.c(2169): [client X.X.X.X:29452] oidc_util_create_symmetric_key: key_len=30
free(): invalid pointer
[Thu Jan 28 12:56:43.625369 2021] [mpm_itk:error] [pid 1410] child died with signal 6
```

**Reply by zandbelt** [reply]
have you set OIDCCryptoPassphrase?

**Reply by wetzel-desy** [reply]
Yes, it is set.

**Reply by zandbelt** [reply]
please paste your config

**Reply by wetzel-desy** [reply]
```
LoadModule auth_openidc_module /usr/lib/apache2/modules/mod_auth_openidc.so

OIDCRedirectURI [redacted]
OIDCCryptoPassphrase [redacted]

# protected locations
<Location /oidc-redirect>
  AuthType openid-connect
  Require valid-user
</Location>

# OAuth2.0 client data
OIDCOAuthClientID [redacted]
OIDCOAuthClientSecret [redacted]
OIDCOAuthVerifyJwksUri https://[redacted]/oauth2/jwk

OIDCOAuthIntrospectionEndpointAuth client_secret_basic

# cache encryption
OIDCCacheEncrypt On
```

**Reply by zandbelt** [reply]
it does not ring a bell but I suspect it has to do with the access token contents itself; if you could send a sample access token and the contents of the `OIDCOAuthVerifyJwksUri` to me in DM I will try and reproduce

**Comment by zandbelt** [comment, marked as answer]
Ok, I reproduced and fixed here: fdbd131

as an OAuth 2.0 client you're not supposed to access paths a that are configured as an OIDC RP (AuthType openid-connect), but you should only access paths configured with AuthType oauth20; if course a crash should prevented, so that is fixed, but basically your path was wrongly configured

**Reply by wetzel-desy** [reply]
Oh, that was quick, thanks! I will try it with the new version and let you know how it turned out.

**Reply by wetzel-desy** [reply]
Now it works again for me. Thanks a lot for the quick action!

**Reply by vlafranca** [reply]
Hi, how to use the new update do I need to compile the package manually ? Not seems to be published in apt registry

**Reply by zandbelt** [reply]
there's no need to install a version that contains the fix to prevent the crash, you can just as wel fix your configuration

**Comment by vlafranca** [comment]
As soon as I activate the module openidc even if not configured for any
virtual host I get a segfault on a particular route which is using guzzle
PHP ...
If I disable openidc all is working fine, any clue ?

Le sam. 30 janv. 2021 à 03:05, Hans Zandbelt <notifications@github.com> a
écrit :
